client/components/tinymce/plugins/simple-payments/dialog/memberships.jsx

```diff
@@ -148,13 +148,13 @@ class MembershipsDialog extends Component {
 		};
 	}
 
-	UNSAFE_componentWillReceiveProps( nextProps ) {
+	componentDidUpdate( prevProps ) {
 		// When transitioning from hidden to visible, show and initialize the form
-		if ( nextProps.showDialog && ! this.props.showDialog ) {
-			if ( nextProps.editPaymentId ) {
+		if ( this.props.showDialog && ! prevProps.showDialog ) {
+			if ( this.props.editPaymentId ) {
 				// Explicitly ordered to edit a particular button
-				this.showButtonForm( nextProps.editPaymentId );
-			} else if ( isEmptyArray( nextProps.paymentButtons ) ) {
+				this.showButtonForm( this.props.editPaymentId );
```

Comment:
here's the loop danger - we're calling `setState()` indirectly inside `componentDidUpdate()`

instead, React created `getDerivedStateFromProps()` to help us infer attributes based on changing state.
this new function returns values to set in state.

```js
getDerivedStateFromProps( props, state ) {
	const openedDialog = props.showDialog && ! state.showDialog;
	const loadedAndEmpty = state.paymentButtons === null && isEmptyArray( props.paymentButtons );

	if ( ! openedDialog && ! loadedAndEmpty ) {
		return null;
	}

	const buttonFormState = editedPaymentId => ( {
		activeTab: 'form',
		editedPaymentId,
		initialFormValues: this.getInitialFormFields( editedPaymentId ),
	} );

	const dialogState = () => {
		if ( props.editPaymentId ) {
			return buttonFormState( props.editPaymentId );
		}

		if ( isEmptyArray( props.paymentButtons ) ) {
			return buttonFormState( null );
		}

		return { activeTab: 'list' };
	}.

	const formState = () => buttonFormState( null );

	return {
		...( openedDialog ? dialogState() : {} ),
		...( loadedAndEmpty ? formState() : {} )
		showDialog: props.showDialog,
		paymentButtons: props.paymentButtons,
	}
}
```

Reply:
I did as you suggested. I prefer to keep helper functions in one file to keep larger pieces of code similar to Simple Payments.
Design hasn't approached this project yet, but when they do - we may want to integrate with Simple Payments modal again and I'd like these codebases to have as few modified functions as possible